### PR TITLE
First working version of the alternative JetMatching for Pythia8.3 and MG5aMCNLO

### PR DESCRIPTION
**Pull request by Cvico**
#### PR description:
This PR includes a plugin that allows to run an alternative Pythia8 FxFx jet matching prescription. The plugin has been implemented using the `PluginFactory` and the code has been directly taken from the new `JetMaching.h` file that can be found [here](http://amcatnlo.web.cern.ch/amcatnlo/FxFx_merging.htm). As stated in the link, this new jet matching algorithm only works in Pythia8.3 and beyond.

New versions of Madgraph5_aMC@NLO (v3.3.0 and beyond) introduce a new clustering procedure that distinguishes between jets that have been generated in a QCD or EWK vertex. This new jet matching prescription is required so the proper treatment of  jets is also accounted for in the parton shower (PS) and matrix element matching step.

An example of its usage can be found in [this fragment example](https://github.com/Cvico/cmssw/blob/JetMatchingEWKFxFx/GeneratorInterface/Pythia8Interface/python/FxFxEWK.py). Basically, the user needs to apply two modifications to the common fragments used for FxFx merging:
* Turn off Pythia8 default matching so new matchings can be add applied UserHooks. This can be done by adding the following line into the process parameters of the fragment:
`'JetMatching:doVeto = off',  # Turn OFF Pythia JetMatching`
* Include the new matching as a UserHook via the UserCustmization method.
```
    UserCustomization = cms.VPSet(
           cms.PSet(pluginName = cms.string("JetMatchingEWKFxFx"))
           ),
```

#### PR validation:

- Tests using the [local config file](https://github.com/Cvico/cmssw/blob/JetMatchingEWKFxFx/GeneratorInterface/Pythia8Interface/python/FxFxEWK.py) show that the JetMatchingEWKFxFx plugin properly matches between Pythia8.3 and MG33X gridpacks.
    - A set of preliminary results for ttW production can be found [here](https://indico.cern.ch/event/1192555/contributions/5086026/attachments/2525008/4342613/ttWnewFxFxSample_CarlosVico_TMG10oct.pdf).
#### Backporting
If merged, we aim for backporting this PR to 12_4_X for Run3 MC production.

### CODE REVIEW

**Comment by SiewYan** [comment]
please test

**Comment by Cvico** [comment]
Dear reviewers,

I was wondering if there's anything more to do from my size for the final inclusion of 
this pull request. I see that code-checks have been ok for quite some time, but I don't
know if there's anything more to do. 

Thank you for your help,
Carlos Vico

**Comment by menglu21** [comment]
please test

**Comment by menglu21** [comment]
+1

File: GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc [hunk 1]
```diff
@@ -0,0 +1,997 @@
...
+
+  // A special version of SlowJet to handle heavy and other partons
+  hjSlowJet = new HJSlowJet(slowJetPower, coneRadius, 0.0, 100.0, 1, 2, nullptr, false, true);
+
+  // Setup local event records
+  eventProcessOrig.init("(eventProcessOrig)", particleDataPtr);
+  eventProcess.init("(eventProcess)", particleDataPtr);
+  workEventJet.init("(workEventJet)", particleDataPtr);
+
+  // Print information
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:133]: `cout` is not thread sale: please think about protecting it with the `MATCHINGDEBUG` directive, as you do in all other cases, or move to some MessageLogger report instead

**Comment by perrotta** [comment]
@Cvico I miss where the `JetMatching:doVeto` flag isread and then used in the code: is it outside your `JetMatchingEWKFxFx` class?

**Comment by perrotta** [comment]
> This PR includes a plugin that allows to run an alternative Pythia8 FxFx jet matching prescription. The plugin has been implemented using the `PluginFactory` and the code has been directly taken from the new `JetMaching.h` file that can be found [here](http://amcatnlo.web.cern.ch/amcatnlo/FxFx_merging.html). As stated in the link, this new jet matching algorithm only works in Pythia8.3 and beyond.

Please notice that the link that you provided does not work: please fix it

**Comment by Cvico** [comment]
Hi @perrotta 

Thank you for your prompt review. For some reason I can't reply to all the comments separately, so I'll address them
in here.

[comment 1.](https://github.com/cms-sw/cmssw/pull/39996#discussion_r1046843443) I guess here you are refering to the Madgraph matching parameters debug text? That piece of code is inherited from the original JetMatching.h file from pythia. Sure I can put it in a matchingdebug block, I don't think that's going to be a problem. I've updated the code according to this suggestion.

[comment 2.](https://github.com/cms-sw/cmssw/pull/39996#issuecomment-1348003053) So as you can see from my config file, I set the JetMatching::doVeto option to OFF in the fragment ([this line](https://github.com/Cvico/cmssw/blob/JetMatchingEWKFxFx/GeneratorInterface/Pythia8Interface/python/FxFxEWK.py#L33)). That tells pythia not to run its own JetMatching algorithm, but rather to expect an external matching algorithm to be called. So that option only affects the JetMatching from within pythia, and not this new matching that I implemented (that's why is not used anywhere in the code effectively)

[comment 3.] (https://github.com/cms-sw/cmssw/pull/39996#issuecomment-1348005385) I've updated the link and now it should be fine. Apologies.

**Comment by Cvico** [comment]
Hi, I'm not entirely sure why the code-checks fail now. I see this in the logs:
gmake[2]: Entering directory '/data/cmsbld/jenkins/workspace/run-pr-code-checks/CMSSW_13_0_X_2022-12-12-2300'

Although this PR was initially for CMSSW_12_6_0_pre4. Is there something I should do? 
Thanks.

**Comment by perrotta** [comment]
@Cvico as the [message](https://github.com/cms-sw/cmssw/pull/39996#issuecomment-1348127177) says with the link below "code format", there is some indentation to fix in your code in order to comply to the cmssw standards.

Please do what suggested in the same [message](https://github.com/cms-sw/cmssw/pull/39996#issuecomment-1348127177), i.e.  run `scram build code-format` in your area, and commit and push the outcome obtained automatically with that `scram` command

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
@cms-sw/generators-l2 is there anything that prevents you reviewing this PR? Could you please comment in any case, telling what you are waiting for, or when do you plan to review?

**Comment by Saptaparna** [comment]
+1

File: GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc [hunk 2]
```diff
@@ -0,0 +1,998 @@
...
+  // Return veto
+  return doVeto;
+}
+
+// Function to get the current number of partons in the Born state, as
+// read from LHE.
+
+int JetMatchingEWKFxFx::npNLO() {
+  string npIn = infoPtr->getEventAttribute("npNLO", true);
+  int np = (npIn.empty()) ? atoi((char*)npIn.c_str()) : -1;
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:992]: One doubt of the last minute: is it correct that if `npIn` is empty you compute `npIn.c_str()`? It does not look so...

In any case `c_str()` returns a `const char*`, which is exaclty the argument needed by `std::atoi`: why do you need to cast a (non const) `char*`?

I think this should be (but please check):
```suggestion
  int np = !npIn.empty() ? atoi(npIn.c_str()) : -1;
```

File: GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc [hunk 3]
```diff
@@ -0,0 +1,998 @@
...
+    if (hjSlowJet->dNext() > qCutSq)
+      break;
+    hjSlowJet->doStep();
+  }
+
+  int nCLjets(0);
+  // Count the number of clusters with pT>qCut.  This includes the
+  //  original hard partons plus any hard emissions.
+  for (int idx = 0; idx < hjSlowJet->sizeAll(); ++idx) {
+    if (hjSlowJet->pT(idx) > sqrt(qCutSq))
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:843]:
```suggestion
    if (hjSlowJet->pT(idx) > qCut)
```
(or is it possible that `qCut` assumes negative values?)

File: GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc [hunk 4]
```diff
@@ -0,0 +1,998 @@
...
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:919]:
```suggestion
    if (hjSlowJet->pT(idx) > qCut)
```
(or is it possible that `qCut` assumes negative values?)

**Comment by rappoccio** [comment]
hold

* Explicitly holding this until comments are addressed.

File: GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc (hunk 3 above)
Comment (Cvico) [reply GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:843]: In principle qCut should not take negative values so indeed this does can be done.

File: GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc (hunk 4 above)
Comment (Cvico) [reply GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc]: In principle qCut should not take negative values so indeed this does can be done.

File: GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc (hunk 3 above)
Comment (perrotta) [reply GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:843]: @Cvico this was not updated with the latest commit...

Comment (perrotta) [reply GeneratorInterface/Pythia8Interface/plugins/JetMatchingEWKFxFx.cc:843]: please commit now

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
unhold

**Comment by Cvico** [comment]
Dear reviewers,

I'd like to thank you for the thorough review of my the plugin, and also to ask you
if there's anything more I can do to proceed with the next steps for the final inclusion
of the PR?

Best,
Carlos Vico

**Comment by perrotta** [comment]
> Dear reviewers,
> 
> I'd like to thank you for the thorough review of my the plugin, and also to ask you if there's anything more I can do to proceed with the next steps for the final inclusion of the PR?
> 
> Best, Carlos Vico

@Cvico just waiting for the @cms-sw/generators-l2 signature (the same as in quite several other PRs which just miss that one...)

**Comment by Saptaparna** [comment]
+1

**Comment by perrotta** [comment]
+1

**Comment by Cvico** [comment]
Hi,
Thank you all for reviewing and finally accepting my PR!

I'd like this to be backported to CMSSW_12_4_X in order to use this in Run3 production. How should
I proceed in order to do this? Thanks in advance for the feedback.

**Comment by perrotta** [comment]
@Cvico please organize with @cms-sw/generators-l2 possible backport strategies of this PR, starting with deciding whether really a production with 12_4_X is what you need now, or if you should point at some more recent release instead

**Pull request merged**